pkg/proxy/iptables/proxier_test.go

```diff
@@ -1345,6 +1347,423 @@ func assertIPTablesRulesNotEqual(t *testing.T, expected, result string) {
+			matches: []string{"--probability 0.5000000000 -j KUBE-SEP-UKSFD7AGPMPPLUHC", "--probability 0.5000000000", "KUBE-SEP-UKSFD7AGPMPPLUHC", "", ""},
+		},
+	}
+
+	for _, testCase := range testCases {
+		t.Run(testCase.name, func(t *testing.T) {
+			matches := testCase.regex.FindStringSubmatch(testCase.rule)
+			if !reflect.DeepEqual(matches, testCase.matches) {
+				t.Errorf("bad match: expected %#v, got %#v", testCase.matches, matches)
+			}
+		})
+	}
+}
+
+var noMatchModules = sets.NewString("conntrack", "mark")
+var knownModules = sets.NewString("addrtype", "recent", "statistic", "tcp")
+
+type iptablesChain []string
+type iptablesTable map[string]iptablesChain
+
+// iptablesTracer holds data used while virtually tracing a packet through a set of
+// iptables rules
+type iptablesTracer struct {
+	tables map[string]iptablesTable
+
+	nodeIP string
+
+	matches  []string
+	outputs  []string
+	markMasq bool
+	markDrop bool
+}
+
+// newIPTablesTracer creates an iptablesTracer. ruleData is an iptables rule dump (as with
+// "iptables-save"). nodeIP is the IP to treat as the local node IP (for determining
+// whether rules with "--src-type LOCAL" or "--dst-type LOCAL" match).
+func newIPTablesTracer(ruleData, nodeIP string) (*iptablesTracer, error) {
+	tables, err := parseIPTablesData(ruleData)
+	if err != nil {
+		return nil, err
+	}
+
+	tracer := &iptablesTracer{
+		tables: make(map[string]iptablesTable),
+		nodeIP: nodeIP,
+	}
+
+	for name, rules := range tables {
+		table := make(iptablesTable)
+		for _, rule := range rules {
+			match := addChainRegex.FindStringSubmatch(rule)
+			if match != nil {
+				table[match[1]] = append(table[match[1]], rule)
+			}
+		}
+		tracer.tables[name] = table
+	}
+
+	return tracer, nil
+}
+
+// ruleMatches checks if the given iptables rule matches (at least probabilistically) a
+// packet with the given sourceIP, destIP, and destPort.
+func (tracer *iptablesTracer) ruleMatches(rule, sourceIP, destIP, destPort string) bool {
+	var match []string
+
+	// Delete comments so we don't mistakenly match something in a comment string later
+	rule = commentRegex.ReplaceAllString(rule, "")
+
+	// Find all of the modules ("-m foo") used; if any of them are in noMatchModules
+	// then the rule is automatically assumed to not match any traffic (for our
+	// purposes). If any of them are not in either noMatchModules or knownModules,
+	// then log a warning and assume the rule doesn't match. (This will only get
+	// triggered if Proxier starts using a new kind of iptables match in the future.)
+	for _, moduleMatches := range moduleRegex.FindAllStringSubmatch(rule, -1) {
+		if noMatchModules.Has(moduleMatches[1]) {
+			return false
+		} else if !knownModules.Has(moduleMatches[1]) {
+			klog.ErrorS(nil, "Rule uses unknown iptables module; assuming no match", "rule", rule, "module", moduleMatches[1])
```

Comment:
I made this panic, though it occured to me later that this isn't _really_ saving us because there are other rules that are allowed without any `-m` flag which we don't make any effort to understand (eg `-i` / `-o`) and for the `knownModules` we don't actually try to parse _all_ of the rules they define (eg, `-m tcp --sport XXX`)